### PR TITLE
add CompareFields

### PR DESCRIPTION
**Pull request by alexzielenski**
/hold
/assign apelisse

adds ability to get a `Comparison` of an object while ignoring certain subpaths using a callback function

### CODE REVIEW

File: typed/typed.go [hunk 1]
```diff
@@ -119,6 +119,10 @@ func (tv TypedValue) Merge(pso *TypedValue) (*TypedValue, error) {
 // match), or an error will be returned. Validation errors will be returned if
 // the objects don't conform to the schema.
 func (tv TypedValue) Compare(rhs *TypedValue) (c *Comparison, err error) {
+	return tv.CompareFields(rhs, nil)
+}
+
+func (tv TypedValue) CompareFields(rhs *TypedValue, callback func(prefix *fieldpath.Path) bool) (c *Comparison, err error) {
```
Comment (apelisse) [inline typed/typed.go:125]: I don't know if I'd make this an arbitrary callback rather than give it a set of fields?

**Comment by apelisse** [comment]
We'll need tests and benchmarks before we can merge that, but go ahead and see if it solves our problem in k/k in the mean time.

File: typed/merge.go [hunk 2]
```diff
@@ -55,6 +60,10 @@ type mergingWalker struct {
 // comparable type.
 type mergeRule func(w *mergingWalker)
 
+// Same as mergeRule but returns a boolean to decide whether the merge/traversal
+// will proceed
+type traversalRule func(w *mergingWalker) bool
```
Comment (lavalamp) [inline typed/merge.go:65]: Probably should call it a predicate, not a rule.

File: typed/merge.go [hunk 3]
```diff
@@ -34,6 +34,11 @@ type mergingWalker struct {
 	// How to merge. Called after schema validation for all leaf fields.
 	rule mergeRule
 
+	// If set, called before both non-leaf and leaf items are merged.
+	// Return `true` if the items should be merged
+	// and `false` if they should not merged or children be traversed
+	preItemHook traversalRule
```
Comment (lavalamp) [inline typed/merge.go:40]: `shouldMerge`?

Comment (lavalamp) [reply typed/merge.go:40]: `proceedOrSkip`?

File: typed/typed.go (hunk 1 above)
Comment (lavalamp) [reply typed/typed.go:125]: Yeah I am inclined to agree that passing in a set of fields to skip is probably best?

**Pull request closed without merging**